### ISSUE TITLE
opensslv.h: New versioning helper macros have multiple issues

### ISSUE THREAD

**Issue by mspncp** [openssl/openssl#5961]
Originally, I just stumbled over a wrong inequality in the OPENSSL_VERSION_AT_LEAST(maj,min) macro. But when attempting to fix it, I noticed that it was only the tip of an iceberg:

Two new macros, OPENSSL_MAKE_VERSION() and OPENSSL_VERSION_AT_LEAST() were proposed by @mcr in pull request #5212, which was merged to master (3c5a61dd0f9d9a9eac098419bcaf47d1c296ca81) resp. 1.1.1 (d8adfdcd2e5de23f3e1d1a1d10c2fda6f4a65c4a). While in general it is a good idea to have such macros which simplify version checks, the aforementioned macros currently not only contain errors, but they also introduce a new interpretation of the versioning scheme, which is problematic because it anticipates a proposal for switching to semantic versioning (issue #5042). Here are the details:

### Wrong inequality in OPENSSL_VERSION_AT_LEAST(maj,min)

The definition of the OPENSSL_VERSION_AT_LEAST(maj,min) macro ([openssl/opensslv.h:48](https://github.com/mcr/openssl/blob/cfcff2ec560d3f5b232df510260116f7ffd80f17/include/openssl/opensslv.h#L48)) is incorrect; the inequality goes in the wrong direction. It should be `<=`, not `>=`:

```C
/* use this for #if tests, should never depend upon fix/patch */
#define OPENSSL_VERSION_AT_LEAST(maj,min) (OPENSSL_MAKE_VERSION(maj,min, 0, 0) >= OPENSSL_VERSION_NUMBER)
```

#### The official scheme

The official versioning scheme is

```
   OpenSSL  <major>.<minor>.<fix><patch_letter>[-status] 
```

which is encoded as in hexadecimal as

```
    0xMmmFFPPS    where   M=<major>  mm=<minor>  FF=<fix> PP=<patch> S=<status>
```
I have written it down here in a slightly different way as in the [documenting comment](https://github.com/mcr/openssl/blob/cfcff2ec560d3f5b232df510260116f7ffd80f17/include/openssl/opensslv.h#L17):

```C
/*-
 * Numeric release version identifier:
 * MNNFFPPS: major minor fix patch status
 * ...
 */
```

#### The semantic interpretation

In [Issue #5042 - Please follow semantic versioning](https://github.com/openssl/openssl/issues/5042#issuecomment-356264737), Richard (@levitte) introduced a new "semantic interpretation" of OpenSSL's versioning scheme which goes as follows:

```
  OpenSSL  1.<major>.<minor><patch_letter>'[-<status>]
```

It seems that the macro OPENSSL_MAKE_VERSION(maj,min,fix,patch) was inspired by this interpretation, see [openssl/opensslv.h:45](https://github.com/mcr/openssl/blob/cfcff2ec560d3f5b232df510260116f7ffd80f17/include/openssl/opensslv.h#L45):

```C
#define OPENSSL_MAKE_VERSION(maj,min,fix,patch) \
    ((0x10000000L)+((maj&0xff)<<20)+((min&0xff)<<12)+((fix&0xff)<<4)+patch)
```

Apart from the error(*) that the lowest nibble is the `status` and not the `patch`, the macro seems to follow Richard's scheme:

```C
#define OPENSSL_MAKE_VERSION(maj,min,patch,status) \
    ((0x10000000L)
        + ((maj&0xff)<<20)
        + ((min&0xff)<<12)
        + ((patch&0xff)<<4)
        + ((status&0xf)<<0)
```

(*) _I would like to emphasize that I don't blame the author of the pull request for this error, but rather the complexity of the versioning scheme. After all, two OMC members reviewed it and didn't notice the glitch._

## The new macros don't simplify the version check

The commit message claims to simplify the version check, but this is not the case: It is not possible to simply write conditional code like the following

```C
#if OPENSSL_VERSION_AT_LEAST(maj,min)
   ...
#endif
```

because the macro will not defined in older letter releases of 1.1.0 and in 1.0.2. Also, it is error prone, because  `OPENSSL_VERSION_AT_LEAST(1,0)`, does _not_ refer to version 1.0, but to version 1.1.0. 

## Proposed Fix

While I am in favour of the proposal for switching to semantic versioning, a change of the versioning scheme can only be done in a major release. To keep the numeric version consistent, at least a version jump from 1.1.x to 1.2.0 will be necessary, which then can be identified with version 2.0 after removing the obsolete "1." prefix. Also, the policies need to be changed, which requires an OMC vote.

The new version macros create precendents, so they should be reverted until we made a decision about further versioning. Because even if one would fix the macros to correctly reflect the current versioning scheme, their presence could become an obstacle to future improvements.

**Comment by levitte** [comment]
Going down memory lane, I can't really say what happened.  (for the unknowing, I was largely away from OpenSSL for some years before 2014, so I may have missed some discussions leading to this).

Like you say, `opensslv.h` does specify major.minor.fix + patch letter.  This change was made for OpenSSL 0.9.3 back in '99 (see commit e90c7729467108ffd6f813ade011c7ec5ced9680).  I cannot remember the discussions from back then (I didn't even have repo access at the time, but do recall participating in the discussions).
I *think* (can't say for sure) that the idea with the patch letters was that these would be for really quick patches, like a few lines of code that fixed something that we would decide should go out to everyone, making a new quick release in the process, or maybe just a diff, like some kind of errata files "attached" to the corresponding full release.  However, our *behaviour* was different, we were never really a group of people that dabbled in quick releases, so if memory serves, the patch letter releases became more involved and further apart in time, so they became fix releases in practice.  I suppose the rest of the scheme simply followed, until it got formalised in a FAQ entry back in 2012 (commit 743fb51d2c821f540304f0c10ab11ed728444875, where the comment in `opensslv.h` was unfortunately not changed to reflect the effective policy)...  and here we are today.

So no, I didn't really introduce a new interpretation in #5042, I only looked at our current practice / policy and noticed that it's similar to the formal semantic versioning scheme, with the twist that it's prefixed with `1.`.

So in summary: our versioning scheme *should* have been much more like the the formal semantic versioning scheme as it was designed in '99, but something happened along the way...

**Comment by mspncp** [comment]
I actually don't want to start a versioning discussion in the current release phase, but leave it for the next major version. Fact is that these new macros currently have bugs and don't work as designed. These bugs either need to be fixed or the commit reverted and the pull request put back on hold until we complete the discussion about future versioning. And we need to do this before that change enters the next official release of 1.1.1 or 1.1.0.

I am in favour of putting that change back on hold. The reason is that I have a simple idea for a smooth transition to semantic versioning for the next major release and I don't want these new macros to create premature facts and obstacles unnecessarily. Even more since their current usefulness is questionable.

I can do the PR for either variant, but I would like to have an indication of your preferences first.

**Comment by mspncp** [comment]
Just in case you are wondering about my ideas for the new versioning scheme: It's mainly a reinterpretation of the current de-facto scheme and does not affect the numerical order of the version numbers.


#### The original (still official) versioning scheme


```
OpenSSL  <major>.<minor>.<fix><patch_letter>[-status] 
```

which is encoded as

```
0xMmmFFPPS    where   M=<major>  mm=<minor>  FF=<fix> PP=<patch> S=<status>
```


#### The de-facto (semantic) versioning scheme


```
OpenSSL  1.<major>.<minor><fix_letter>[-status] 
```

which is encoded as

```
0x1MMmmFFS    where   M=<major>  mm=<minor>  FF=<fix> S=<status>
```

#### The new (semantic) versioning scheme

The idea behind the new scheme, beginning with OpenSSL 1.2.0, is to drop the defunct "1." prefix, i.e., go right to OpenSSL 2.0 without recoding but only reinterpreting the release number:


```
OpenSSL  <major>.<minor>.<fix>[-status] 
```

which is encoded as

```
0x8MMmmFFS    where   M=<major>  mm=<minor>  FF=<fix> S=<status>
```

**Note:** The first nibble was changed from 0x1 to 0x8 to indicate a change of the versioning scheme. I did not just increment it to 0x2, because it then easily could have got mixed up with the new major version number. Instead, it just setst he topmost bit and leaves three bits unused. This change leaves the numerical order of the version numbers intact.

**Comment by mspncp** [comment]
_After_ such a change is decided, one can (re-)introduce the macro OPENSSL_VERSION_AT_LEAST(major,minor) and backport it to all stable versions, including 1.0.2 (without changing the macro, since only the interpretation has changed). Only if the macro is available on all supported branches can it be generally useful.

**Comment by levitte** [comment]
A small correction...  what you call the "de-facto" versioning scheme is actually the official one, at the very least since 2012.  Therefore, the original from '99 is *not* the current official one.  For reference, see our [release strategy](https://www.openssl.org/policies/releasestrat.html), which essentially says the same thing as our FAQ [has said for a while](https://www.openssl.org/docs/faq.html#MISC7).

Now personally, I don't really care that much about `OPENSSL_VERSION_AT_LEAST`, so if people find it more confusing than helpful, I agree that it should go.  And, *whatever* we end up deciding, the bugs in it (oops) should of course be corrected.  Were you thinking of creating a PR for it?

**Comment by levitte** [comment]
Oh, and the aged comment in `opensslv.h` should of course be updated to correspond to today's reality.

**Comment by mspncp** [comment]
> Were you thinking of creating a PR for it?

PR for reverting the change is in #5968.

> Oh, and the aged comment in opensslv.h should of course be updated to correspond to today's reality.

Since it's been wrong for such a long time, correction can wait until we decided whether we switch to the new semantic versioning scheme for OpenSSL 2.0  (1.2.0 in the old notation) or not.

**Comment by mcr** [comment]
If you remove the macro, then downstream libraries (such as ruby-openssl)
will be decoding the version integer directly, and then the version number
formats will be essentially set in stone.   I didn't propose this patch for
fun, but because it removed interpretation of the version number from
downstream code bases.

**Comment by mcr** [comment]
Matthias St. Pierre <notifications@github.com> wrote:
    > release phase, but leave it for the next major version. Fact is that
    > these new macros currently have bugs and don't work as designed. These

Sure, maybe you could start by explaining the bugs?

**Comment by mspncp** [comment]
I did, in the description of this issue.

**Comment by mspncp** [comment]
Error one is the wrong equality. Error two was the encoding of the last nibble, the status field. I even showed the correction:

```C
#define OPENSSL_MAKE_VERSION(maj,min,patch,status) \
    ((0x10000000L)
        + ((maj&0xff)<<20)
        + ((min&0xff)<<12)
        + ((patch&0xff)<<4)
        + ((status&0xf)<<0)
```

**Comment by mspncp** [comment]
See also https://github.com/openssl/openssl/pull/5968#issuecomment-381625989.

**Comment by mcr** [comment]
I can't understand your complaint about the test being the wrong direction.

    /* use this for #if tests, should never depend upon fix/patch */
    #define OPENSSL_VERSION_AT_LEAST(maj,min) (OPENSSL_MAKE_VERSION(maj,min, 0, 0) >= OPENSSL_VERSION_NUMBER)

I went back to the source code on my laptop to make sure that I'm looking at the code in the tree.
First, we make a version using MAKE_VERSION().  
Then we see if that generated number is bigger (or equal) to OPENSSL_VERSION_NUMBER.  How is the >= wrong?

As for the macro not being available in previous versions, that's indeed a problem. If the macro does not exist, then one must assume an earlier version.  In many cases, that's enough to know a 1.1 feature is not available.

**Comment by mspncp** [comment]
> If you remove the macro, then downstream libraries (such as ruby-openssl)
will be decoding the version integer directly, and then the version number
formats will be essentially set in stone.   I didn't propose this patch for
fun, but because it removed interpretation of the version number from
downstream code bases.

Could you please point out where and how you are currently using this macro (e.g. a link into your repository)?

**Comment by mspncp** [comment]
My interpretation of its use is: The application includes `<openssl/opensslv.h>` and then checks whether the OpenSSL version in the included header is at least (a.k.a greater or equal than) the given major and minor version. Mathematically, that is `OPENSSL_VERSION_NUMBER >= OPENSSL_MAKE_VERSION(maj,min, 0, 0)`, right?

**Comment by mspncp** [comment]
> Back it out and do it whatever way you think right.

Please don't get me wrong. I think your macro is a good idea and it should be fixed and added to all stable branches. I'm just not sure whether this is the right moment. And it needs to be documented properly. Also, I would prefer if we could introduce the macros together with a 'correct' semantic version numbering. Then your macros would fit more naturally.

But if the others prefer it, I could also correct the macro right away and document it properly. In that case it would make sense to backport the fixed macro to 1.0.2, too. @levitte and @mattcaswell, your opinion? Should I correct it now or postpone it?

> I never got to using it in ruby-openssl, because I found another way to deal, since I had to deal with old versions at this time: I check the library for the new function I care about.

Checking for feature presence instead of version numbers is always the better choice, if available.

**Comment by mspncp** [comment]
@richsalz just approved the removal (#5968). So if no one objects, I would suggest that we postpone the discussion to post-1.1.1 and I'll prepare a pull request for a corrected implementation. @mcr I will ping you when it's ready.

**Comment by mattcaswell** [comment]
>  @mattcaswell, your opinion? 

Remove it.
